File: app/src/main/java/com/soundcloud/android/discovery/recommendedplaylists/RecommendedPlaylistsOperations.java
package com.soundcloud.android.discovery.recommendedplaylists;

import com.soundcloud.android.discovery.DiscoveryItem;
import com.soundcloud.android.model.Urn;
import com.soundcloud.android.playlists.Playlist;
import com.soundcloud.android.playlists.PlaylistItem;
import com.soundcloud.android.playlists.PlaylistRepository;
import com.soundcloud.android.presentation.EntityItemCreator;
import com.soundcloud.android.sync.SyncOperations;
import com.soundcloud.android.sync.Syncable;
import rx.Observable;
import rx.functions.Func1;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class RecommendedPlaylistsOperations {

    private final SyncOperations syncOperations;
    private final RecommendedPlaylistsStorage storage;
    private final PlaylistRepository playlistRepository;
    private final EntityItemCreator entityItemCreator;

    @Inject
    RecommendedPlaylistsOperations(SyncOperations syncOperations,
                                   RecommendedPlaylistsStorage storage,
                                   PlaylistRepository playlistRepository,
                                   EntityItemCreator entityItemCreator) {
        this.syncOperations = syncOperations;
        this.storage = storage;
        this.playlistRepository = playlistRepository;
        this.entityItemCreator = entityItemCreator;
    }

    public Observable<DiscoveryItem> recommendedPlaylists() {
        return syncOperations.lazySyncIfStale(Syncable.RECOMMENDED_PLAYLISTS)
                             .flatMap(o -> readRecommendedPlaylistsFromStorage());
    }

    public Observable<DiscoveryItem> refreshRecommendedPlaylists() {
        return syncOperations.failSafeSync(Syncable.RECOMMENDED_PLAYLISTS)
                             .flatMap(o -> readRecommendedPlaylistsFromStorage());
    }

    private Observable<RecommendedPlaylistsBucketItem> fromEntities(final List<RecommendedPlaylistsEntity> buckets) {
        final Set<Urn> recommendedPlaylists = new HashSet<>();
        for (RecommendedPlaylistsEntity bucket : buckets) {
            recommendedPlaylists.addAll(bucket.playlistUrns());
        }

        if (recommendedPlaylists.isEmpty()) {
            return Observable.empty();
        }

        return playlistRepository.withUrns(recommendedPlaylists)
                                 .flatMap(toOrderedBuckets(buckets));
    }

    private Func1<Map<Urn, Playlist>, Observable<? extends RecommendedPlaylistsBucketItem>> toOrderedBuckets(List<RecommendedPlaylistsEntity> buckets) {
        return playlistEntities -> Observable
                .from(buckets)
                .map(mapToBucketItem(playlistEntities));
    }

    private Func1<RecommendedPlaylistsEntity, RecommendedPlaylistsBucketItem> mapToBucketItem(final Map<Urn, Playlist> playlistEntities) {
        return entity -> {
            final List<PlaylistItem> matches = new ArrayList<>(entity.playlistUrns().size());
            for (Urn urn : entity.playlistUrns()) {
                matches.add(entityItemCreator.playlistItem(playlistEntities.get(urn)));
            }
            return RecommendedPlaylistsBucketItem.create(entity, matches);
        };
    }

    private Observable<? extends DiscoveryItem> readRecommendedPlaylistsFromStorage() {
        return storage.recommendedPlaylists()
                      .flatMap(new Func1<List<RecommendedPlaylistsEntity>, Observable<RecommendedPlaylistsBucketItem>>() {
                          @Override
                          public Observable<RecommendedPlaylistsBucketItem> call(List<RecommendedPlaylistsEntity> entities) {
                              return fromEntities(entities);
                          }
                      });
    }

}

File: app/src/main/java/com/soundcloud/android/olddiscovery/recommendedplaylists/RecommendedPlaylistsOperations.kt
package com.soundcloud.android.olddiscovery.recommendedplaylists

import com.soundcloud.android.model.Urn
import com.soundcloud.android.olddiscovery.OldDiscoveryItem
import com.soundcloud.android.playlists.Playlist
import com.soundcloud.android.playlists.PlaylistRepository
import com.soundcloud.android.presentation.EntityItemCreator
import com.soundcloud.android.sync.NewSyncOperations
import com.soundcloud.android.sync.Syncable
import com.soundcloud.android.utils.OpenForTesting
import io.reactivex.Observable
import io.reactivex.functions.Function
import javax.inject.Inject

@OpenForTesting
class RecommendedPlaylistsOperations
@Inject
internal constructor(private val syncOperations: NewSyncOperations,
                     private val storage: RecommendedPlaylistsStorage,
                     private val playlistRepository: PlaylistRepository,
                     private val entityItemCreator: EntityItemCreator) {

    fun recommendedPlaylists(): Observable<OldDiscoveryItem> {
        return syncOperations.lazySyncIfStale(Syncable.RECOMMENDED_PLAYLISTS)
                .flatMapObservable { _ -> readRecommendedPlaylistsFromStorage() }

    }

    fun refreshRecommendedPlaylists(): Observable<OldDiscoveryItem> {
        return syncOperations.failSafeSync(Syncable.RECOMMENDED_PLAYLISTS)
                .flatMapObservable { _ -> readRecommendedPlaylistsFromStorage() }
    }

    private fun readRecommendedPlaylistsFromStorage(): Observable<OldDiscoveryItem> {
        return storage.recommendedPlaylists()
                .toObservable()
                .flatMap { this.fromEntities(it) }
    }

    private fun fromEntities(buckets: List<RecommendedPlaylistsEntity>): Observable<RecommendedPlaylistsBucketItem> {
        val recommendedPlaylists = buckets.flatMap{ it.playlistUrns }

        if (recommendedPlaylists.isEmpty()) {
            return Observable.empty()
        }

        return playlistRepository.withUrns(recommendedPlaylists)
                .flatMapObservable(toOrderedBuckets(buckets))
    }

    private fun toOrderedBuckets(buckets: List<RecommendedPlaylistsEntity>): Function<Map<Urn, Playlist>, Observable<out RecommendedPlaylistsBucketItem>> {
        return Function {
            Observable
                    .fromIterable(buckets)
                    .map(mapToBucketItem(it))
        }
    }

    private fun mapToBucketItem(playlistEntities: Map<Urn, Playlist>): Function<RecommendedPlaylistsEntity, RecommendedPlaylistsBucketItem> {
        return Function {
            val matches = it.playlistUrns.map { entityItemCreator.playlistItem(playlistEntities[it]) }
            RecommendedPlaylistsBucketItem.create(it, matches)
        }
    }
}
